lldb/source/Plugins/ExpressionParser/Swift/SwiftUserExpression.cpp

```diff
@@ -144,9 +144,13 @@ findSwiftSelf(StackFrame &frame, lldb::VariableSP self_var_sp) {
     info.is_metatype = true;
   }
 
+  info.swift_type = GetSwiftType(info.type).getPointer();
+  if (auto dyn_self =
+      llvm::dyn_cast_or_null<swift::DynamicSelfType>(info.swift_type))
+    info.swift_type = dyn_self->getSelfType().getPointer();
+
   // 5) If the adjusted type isn't equal to the type according to the runtime,
   // switch it to the latter type.
-  info.swift_type = GetSwiftType(info.type).getPointer();
```

Comment:
> Why doesn't `BindGenericTypeParameters`, which must also be called nearby, suffice?

On first quick look, I don't see where `BindGenericTypeParameters` is called. It's not called in any of `SwiftASTManipulator.cpp`, `SwiftExpressionSourceCode.cpp`, or `SwiftUserExpression.cpp`.

Reply:
>> Why doesn't BindGenericTypeParameters, which must also be called nearby, suffice?
> On first quick look, I don't see where BindGenericTypeParameters is called nearby. It's not called in any of SwiftASTManipulator.cpp, SwiftExpressionSourceCode.cpp, or SwiftUserExpression.cpp.

You are right, nearby is a stretch: it's called in `AddRequiredAliases()` and `GetSwiftTypeForVariableValueObject()`, which are called by the LLDB callbacksthat provide unresolved types in the expression to the Swift compiler. Which, I assume, is too late to affect the if-cascade with the `mutating` keyword.

Reply:
I don't know what the correct long term decision is, but there are some unexpected results happening beyond the core `mutating` keyword issue.

First, in the `convenience init`, nothing is explicitly declared using `Self`. It was a surprise to me that it would automatically take on a type of `Self`.

Second, compare the two initializers in this code:

```swift
class C {
  init() {}
  convenience init(unused: Bool) { self.init() }
}
```

The types of `self` are different in the designated initializer (`main.C`) vs the convenience initializer (`Self`), which I would not expect. 

`init()`

```
(lldb) v self
(main.C) self = 0x0000000100504080 {}
```

`convenience init(unused:)`

```
(lldb) v self
Self self =
(lldb) v -d run-target self
Self self =
```

Reply:
I think the patch is fine.